tidb-server/server/conn.go

```diff
@@ -0,0 +1,469 @@
+
+type clientConn struct {
+	pkg          *packetIO
+	conn         net.Conn
+	server       *Server
+	capability   uint32
+	connectionID uint32
+	collation    uint8
+	charset      string
+	user         string
+	dbname       string
+	salt         []byte
+	alloc        arena.Allocator
+	lastCmd      string
+	ctx          IContext
+}
+
+func (cc *clientConn) String() string {
+	return fmt.Sprintf("conn: %s, status: %d, charset: %s, user: %s, lastInsertId: %d",
+		cc.conn.RemoteAddr(), cc.ctx.Status(), cc.charset, cc.user, cc.ctx.LastInsertID(),
+	)
+}
+
+func (cc *clientConn) handshake() error {
+	if err := cc.writeInitialHandshake(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := cc.readHandshakeResponse(); err != nil {
+		cc.writeError(err)
+		return errors.Trace(err)
+	}
+	data := cc.alloc.AllocWithLen(4, 32)
+	data = append(data, mysql.OKHeader)
+	data = append(data, 0, 0)
+	if cc.capability&mysql.ClientProtocol41 > 0 {
+		data = append(data, dumpUint16(mysql.ServerStatusAutocommit)...)
+		data = append(data, 0, 0)
+	}
+
+	err := cc.writePacket(data)
+	cc.pkg.sequence = 0
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(cc.flush())
+}
+
+func (cc *clientConn) Close() error {
+	cc.server.rwlock.Lock()
+	delete(cc.server.clients, cc.connectionID)
+	cc.server.rwlock.Unlock()
+	cc.conn.Close()
+	return cc.ctx.Close()
+}
+
+func (cc *clientConn) writeInitialHandshake() error {
+	data := make([]byte, 4, 128)
+
+	//min version 10
+	data = append(data, 10)
+	//server version[00]
+	data = append(data, mysql.ServerVersion...)
+	data = append(data, 0)
+	//connection id
+	data = append(data, byte(cc.connectionID), byte(cc.connectionID>>8), byte(cc.connectionID>>16), byte(cc.connectionID>>24))
+	//auth-plugin-data-part-1
+	data = append(data, cc.salt[0:8]...)
+	//filter [00]
+	data = append(data, 0)
+	//capability flag lower 2 bytes, using default capability here
+	data = append(data, byte(defaultCapability), byte(defaultCapability>>8))
+	//charset, utf-8 default
+	data = append(data, uint8(mysql.DefaultCollationID))
+	//status
+	data = append(data, dumpUint16(mysql.ServerStatusAutocommit)...)
+	//below 13 byte may not be used
+	//capability flag upper 2 bytes, using default capability here
+	data = append(data, byte(defaultCapability>>16), byte(defaultCapability>>24))
+	//filter [0x15], for wireshark dump, value is 0x15
+	data = append(data, 0x15)
+	//reserved 10 [00]
+	data = append(data, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
+	//auth-plugin-data-part-2
+	data = append(data, cc.salt[8:]...)
+	//filter [00]
+	data = append(data, 0)
+	err := cc.writePacket(data)
+	if err != nil {
+		return err
+	}
+	return cc.flush()
+}
+
+func (cc *clientConn) readPacket() ([]byte, error) {
+	return cc.pkg.readPacket()
+}
+
+func (cc *clientConn) writePacket(data []byte) error {
+	return cc.pkg.writePacket(data)
+}
+
+func calcPassword(scramble, password []byte) []byte {
+	if len(password) == 0 {
+		return nil
```

Comment:
It is used by `bytes.Equal` and according to the comment `A nil argument is equivalent to an empty slice.`, it doesn't matter.